### PR TITLE
Shader improvements: Saturation Optimization

### PR DESCRIPTION
**Pull request by Pevernow**
#####

Saturation Optimization OFF:

原版色调映射

Saturation Optimization ON:
饱和度优化

## What it did?
This PR changes the saturation of the blocks in the rendering.

Make the color of minetest brighter.

It can effectively improve the image quality without any performance loss

Through repeated tests, it is found that the saturation parameter is 1.5, which looks best.

Because it is only a simple change of tone, it is put together with tone mapping.

(when hue mapping is enabled in minetest, it is also enabled)

## How to test

1.Start minetest before merging and take a screenshot after entering the world

2.Then start minetest in this PR and take a screenshot after entering the world

3.Compare these two screenshots

### CODE REVIEW

**Comment by rollerozxa** [comment]
so now when tone mapping is enabled it will become even more of a brightly colored eyesore?

**Comment by Pevernow** [comment]
Yes, it improves the tone of minetest.

**Comment by TurkeyMcMac** [comment]
Would it be possible to give users control over saturation in the settings page?

**Comment by LoneWolfHT** [comment]
This looks nice, being able to control the amount via a setting is a must though

**Comment by lhofhansl** [comment]
Looks nice indeed. If possible let's avoid another config option. We have too many of these already.

**Comment by LoneWolfHT** [comment]
> Looks nice indeed. If possible let's avoid another config option. We have too many of these already.

I'd say it's better than needing to edit & compile a custom client/modify the shader code because you want a little less saturation

**Comment by NathanSalapat** [comment]
In some games this could look nice, but if it's supposed to be dark and gritty this would break immersion. It would have to be disable-able by a server.

**Comment by nerzhul** [comment]
it's a nice shader improvement. If we can handle saturation factor in a setting it would be better

**Comment by Pevernow** [comment]
> In some games this could look nice, but if it's supposed to be dark and gritty this would break immersion. It would have to be disable-able by a server.

@NathanSalapat @sfan5 
So should I add this in the client settings or in the server settings?

**Comment by Pevernow** [comment]
Saturation parameters can now be adjusted in settings.

**Comment by Pevernow** [comment]
@sfan5 @nerzhul Is this OK now?

**Comment by lhofhansl** [comment]
Looks good.
(I expect some "whining" about the changed default, though :) )

**Comment by Pevernow** [comment]
@SmallJoker This PR looks ready to merge.

**Comment by lhofhansl** [comment]
@Pevernow Just needs a second approval.

**Comment by rubenwardy** [comment]
(Supported by lhofhansl)

**Comment by Pevernow** [comment]
@sfan5 @SmallJoker 
This pull request has been ignored for a long time.
Could you approve this pull request,please?

**Comment by rubenwardy** [comment]
@x2048

**Comment by lhofhansl** [comment]
Yeah. Maybe leave the default at 1.0, and call it just "saturation".

Also, since this is great to invoke certain moods in a game it would be *awesome* if this could also be controlled by Lua on the server.

@Pevernow do you time (and willingness) to make this server controlled as well?

You can see #11944 about how to do that. In fact, clouds, sky color, shadows, saturation, etc, all go extremely well together.
For example I've been trying out some de-saturation, and it really is helpful for some kind of games, together with shadows and cloud cover.

**Comment by x2048** [comment]
I agree with @rubenwardy, 1.5 is a default is strong, and in general perception of saturation is quite subjective, and there is no 'better' or 'worse' values.

I'd suggest going with multipliers from both game's/server's and player's side. The game/server can dynamically change saturation to reflect the mood/situation, and player can adjust the scale of saturation for their hardware/texture pack.

**Comment by Pevernow** [comment]
> @x2048
> 
> I think the default effect is way too strong - I prefer the before to the after.
> 
> The name "saturation optimization" seems weird to me - why does it effect brightness? Saturation shouldn't effect value

So, what should I set the default value to?

1.5？ 1.0？ Or something else?

**Comment by Pevernow** [comment]
> Yeah. Maybe leave the default at 1.0, and call it just "saturation".
> 
> Also, since this is great to invoke certain moods in a game it would be _awesome_ if this could also be controlled by Lua on the server.
> 
> @Pevernow do you have time (and willingness) to make this server controlled as well?
> 
> You can see #11944 about how to do that. In fact, clouds, sky color, shadows, saturation, etc, all go extremely well together. For example I've been trying out some de-saturation, and it really is helpful for some kind of games, together with shadows and cloud cover.

I'm afraid I don't have enough time to complete the API.

Moreover, completing this function at the same time as @x2048's pull request will lead to certain function overlap, and we have to spend more time dealing with the conflict between the two pull requests

I would like to hand over the task of completing API to @x2048.

@x2048 Could you help me finish it?

**Comment by x2048** [comment]
@Pevernow  I will create the base API and networking code in #11944 and can help you integrate when it's finalized. 

You would help me if you change the default to 1.0 and add another uniform as a multiplier, keeping at a constant 1.0. I will then hook it to the server API. I propose the names `base_saturation` (from local settings) and `dynamic_saturation` (from API) for the uniforms. This will lay the ground for mutual control by the game and the player. Thank you.

**Comment by Zughy** [comment]
Please don't alter the basic value. MT basic texture pack is quite dull, but improving it through altering the overall saturation of the engine is not a healthy approach. Imagine something like [Soothing 32](https://content.minetest.net/packages/Zughy/soothing32/) (the first texture pack on CDB) with something like this: contrast will simply kill you. Texture packs in general have all been designed to fit the current MT standard, applying a new one (1.5) would mean "breaking" all of them

**Comment by x2048** [comment]
@Zughy There is no intention to go forward with 1.5.

**Comment by lhofhansl** [comment]
I'm also happy to add the API (time permitting),

**Comment by Pevernow** [comment]
The stable version of 5.5.0 has been released.

Would you like to merge this now?

File: client/shaders/nodes_shader/opengl_fragment.glsl
Comment (appgurueu) [inline client/shaders/nodes_shader/opengl_fragment.glsl:463]: Unneded comment

**Review by appgurueu — CHANGES_REQUESTED** [review]
I don't like that the setting is called `_param`. Call it `factor` or something meaningful, but `param` is just clutter.

**Comment by x2048** [comment]
@Pevernow I remember we were discussing adding and API and at least changing the default to 1.0 before merging this PR.

@lhofhansl and I can help wiring this to the network and Lua if you expose a C++ method to control the saturation on the client side.

**Review by Df458 — CHANGES_REQUESTED** [review]
Tested, and frankly I don't think this is in a state we should merge.
Obviously, there's the contention over defaults and API exposure. I'm not too concerned about these though, since adjusting the default is pretty easy to do and I see no reason why we can't add an API to extend the feature after merging this.

However, the problem is much simpler: This shader adjustment doesn't effect enough. As OP mentioned in passing, it changes the saturation of **nodes**. Not entities, not the sky, _just nodes_. This causes some problems, which are easiest to demonstrate at a saturation of 0:
Screenshot-2022-02-16-07:43:13

Missing support pictured in this collage:
- Falling nodes
- Dig particles
- Nodes with an entity component for animated/dynamic mesh visuals
- Wielded nodes/items
- **The sky**

...the list goes on. I don't think we should be accepting a visual adjustment like this, which just casually skips large amounts of visuals. In its current state, this PR will only exacerbate the existing visual inconsistencies in much of MT's content. If you want this in, please
1. Change the default to 1. From a quick glance (though I'll admit, I didn't count), it looks like the majority of the coredevs and players who've chimed in would like the default unchanged and I'm also in this camp.
2. Make the application of saturation to the world as consistent as possible. If you can't do it to UI elements for whatever reason, that's one thing. But I think supporting the sky and entities is the bare minimum, and supporting particle effects is also rather important.

**Comment by Df458** [comment]
Also, please rebase so it's easier to test. The code itself doesn't need a rebase, but it's no longer compatible with the current version of IrrlictMT so it'd be nice if you could update it on your end.

**Comment by appgurueu** [comment]
Another issue: This is not implemented in a post-processing stage. As such it's bound to be rather inefficient, slowing down the rendering of everything as it will have to change the saturation of every fragment, including occluded fragments.

**Comment by lhofhansl** [comment]
@Df458 Yeah, agree that should affect the object shader as well.
@appgurueu Sure, but we do not have a post-processing pipeline - and will likely never have one, see all previous efforts. And it does not add anything I could measure over just tone mapping.

I've been using this with saturation set to 0.7 and it changes the entire game for me.
If this works for entities, which is not hard to do, I'm still for merging this (even without server control... Gamma et al are also client settings).
Is tonemapping implemented in the object shader? (Yeah could just check) If not we have another problem as well.

**Comment by rubenwardy** [comment]
Tonemapping should absolutely be done in post processing, it's a post processing effect.  I don't know what "saturation optimization" is supposed to be, but if it's just saturation color mapping then that sounds a lot like a post processing effect too

> Sure, but we do not have a post-processing pipeline - and will likely never have one, see all previous efforts

There is absolutely interest in adding a post-processing pipeline. Hell, even paramat supported adding post processing. It's the better way of doing things like tonemapping

> And it does not add anything I could measure over just tone mapping.

Post-processing can be used for other things, like depth of field, bloom(?), and probably a lot more. That being said, tonemapping by itself is enough of a justification imo to move to post processing

**Comment by lhofhansl** [comment]
Thanks @rubenwardy .
Just for the record, I'm a big fan of added post processoing. #5446 was filed 5 years ago, though.

I was about to pick this up and make it part of the lighting API. We can argue whether this is part of the lighting API, I think it is, saturation in the real world is different when the sun is shining vs. cloudy; can also invent to new API.

So this one... Just wait for now?

**Comment by appgurueu** [comment]
> I was about to pick this up and make it part of the lighting API. We can argue whether this is part of the lighting API, I think it is, saturation in the real world is different when the sun is shining vs. cloudy; can also invent to new API.

It definitely is ambient lighting.

**Comment by lhofhansl** [comment]
Lemme me create a followup PR when I get time. It will use the lighting API and then we can have a discussion on the specific merits.

As with shadows I would strongly prefer a server "don't care" option which should be default... We'll see.

File: src/defaultsettings.cpp
Comment (appgurueu) [inline src/defaultsettings.cpp:259]: Missing a space after the comma.

File: src/client/shader.cpp
Comment (appgurueu) [inline src/client/shader.cpp:733]: The `_PARAM` suffix is superfluous. Not complying with formatting rules (missing a space).

File: minetest.conf.example
Comment (appgurueu) [inline minetest.conf.example:634]: Space after `min:`

File: client/shaders/nodes_shader/opengl_fragment.glsl
Comment (appgurueu) [inline client/shaders/nodes_shader/opengl_fragment.glsl:464]: should be called `saturate` or the like

**Comment by rubenwardy** [comment]
This was discussed in the last meeting and today's meeting. We've decided to close this for a number of reasons, the biggest (for me) is https://github.com/minetest/minetest/pull/11854#issuecomment-1048403113

**Comment by lhofhansl** [comment]
We have post processing now.

Saturation can greatly affect game mood. I.e. de-saturation when it's cloudy or rainy, black-white when you are sick, colorful on islands, etc, etc... Would obviously have to be server controlled.

@x2048 What's the best way to add this to the current framework?

I'm re-opening this to have a discussion.

**Comment by HybridDog** [comment]
Minetest does not use spectral rendering, so a component-wise multiplication of RGB vectors for diffuse lighting is physically inaccurate (see for example https://mitsuba.readthedocs.io/en/latest/src/key_topics/variants.html#part-3-color-representation).
I think that instead of using a vector `s` for the sunlight colour, we could use a 3x3 matrix `M`. The result of a diffuse reflection would then be `M c` instead of `diag(s) c` for a texel colour `c`. This would enable more degrees of freedom, e.g. saturation control, slight hue shifts, etc. However, I think it requires that the colours are in linearised sRGB.

**Pull request closed without merging**
